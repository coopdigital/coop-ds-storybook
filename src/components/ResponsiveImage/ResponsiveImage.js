import React from "react";
import PropTypes from "prop-types";
import shortid from "shortid";
import classNames from "../../utils/classNames";
import "./responsiveImage.scss";

const generateSourceMap = (sources) =>
  sources.map((source) => {
    const { type, media, srcSet } = source;
    const srcSets = Object.entries(srcSet).reduce(
      (acc, [descriptor, imgSrc]) => {
        acc.push(`${imgSrc} ${descriptor}`);
        return acc;
      },
      []
    );

    const sourceProps = { type, media };
    return (
      <source
        key={shortid.generate()}
        srcSet={srcSets.join(", ")}
        {...sourceProps}
      />
    );
  });

const ResponsiveImage = ({
  src,
  alt,
  width,
  height,
  sources,
  className,
  figureClass,
  pictureClass,
  imageClass }) => {

  const figureClasses = classNames("coop-c-image", [
    figureClass && figureClass,
    className,
  ]);
  const pictureClasses = classNames("coop-c-image__media", [
    pictureClass && pictureClass,
    className,
  ]);
  const imageClasses = classNames(null, [
    imageClass && imageClass,
    className,
  ]);

  return (
    <figure className={figureClasses}>
      <picture className={pictureClasses}>
        {generateSourceMap(sources)}
        <img
          className={imageClasses}
          src={src}
          alt={alt}
          width={width}
          height={height} />
      </picture>
    </figure>
  );
};

ResponsiveImage.defaultProps = {
  width: null,
  height: null,
  sources: [],
};

ResponsiveImage.propTypes = {
  /** src - (required) a url to the image you are sourcing. */
  src: PropTypes.string.isRequired,
  /** alt - (required) needed to give a meaningful description to screenreader users. */
  alt: PropTypes.string.isRequired,
  /** width - can set the width of the image in pixels */
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** height - can set the height of the image in pixels */
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** source - takes an array of objects with the following properties: `type`, `media` and `srcSet` */
  sources: PropTypes.arrayOf(PropTypes.object),
  /** figureClass - allows you to override|add additional classes to target the `<figure>` tag within the component  */
  figureClass: PropTypes.string,
  /** pictureClass - allows you to override|add additional classes to target the `<picture>` tag within the component */
  pictureClass: PropTypes.string,
  /** imageClass - - allows you to override|add additional classes to target the `<img>` tag within the component */
  imageClass: PropTypes.string
};

export default ResponsiveImage;
