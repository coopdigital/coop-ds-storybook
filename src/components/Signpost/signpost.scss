@import "@coopdigital/shared-component--signpost/dist/signpost.css";
