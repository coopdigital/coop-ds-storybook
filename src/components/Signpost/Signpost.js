import React from "react";
import PropTypes from "prop-types";
import classNames from "../../utils/classNames";
import "./signpost.scss";

const Signpost = ({ title, url, mediaComponent }) => {
  const rootClasses = classNames("coop-c-signpost", [
    mediaComponent && "coop-c-hero--media",
  ]);

  return (
    <>
      <div className={rootClasses}>
        <a
          href={url}
          class="coop-c-signpost__link"
          data-contenttype=""
          data-contentparent=""
          data-linktext=""
        >
          {mediaComponent && (
            <figure class="coop-c-signpost__media">
              <picture class="coop-c-signpost__image">
                <source
                  media="(min-width: 48em)"
                  srcset="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=webp&amp;fit=thumb&amp;q=60&amp;w=360&amp;h=203 1x,
        //images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=webp&amp;fit=thumb&amp;q=60&amp;w=720&amp;h=406 2x"
                  type="image/webp"
                />
                <source
                  media="(min-width: 37.5em)"
                  srcset="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=webp&amp;fit=thumb&amp;q=60&amp;w=751&amp;h=423 1x,
        //images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=webp&amp;fit=thumb&amp;q=60&amp;w=1502&amp;h=846 2x"
                  type="image/webp"
                />
                <source
                  media="(min-width: 25.9375em)"
                  srcset="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=webp&amp;fit=thumb&amp;q=60&amp;w=583&amp;h=329 1x,
        //images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=webp&amp;fit=thumb&amp;q=60&amp;w=1166&amp;h=658 2x"
                  type="image/webp"
                />
                <source
                  media="(min-width: 48em)"
                  srcset="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;fl=progressive&amp;q=60&amp;w=360&amp;h=203 1x,
        //images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;fl=progressive&amp;q=60&amp;w=720&amp;h=406 2x"
                  type="image/jpeg"
                />
                <source
                  media="(min-width: 37.5em)"
                  srcset="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;fl=progressive&amp;q=60&amp;w=751&amp;h=423 1x,
        //images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;fl=progressive&amp;q=60&amp;w=1502&amp;h=846 2x"
                  type="image/jpeg"
                />
                <source
                  media="(min-width: 25.9375em)"
                  srcset="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;fl=progressive&amp;q=60&amp;w=583&amp;h=329 1x,
        //images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;fl=progressive&amp;q=60&amp;w=1166&amp;h=658 2x"
                  type="image/jpeg"
                />
                <img
                  src="//images.ctfassets.net/bffxiku554r1/4wuuKMjqUbsIJL7gl65A7I/3ebd189f497edaf54091615e67203910/Memb.png?fm=jpg&amp;fit=thumb&amp;q=60&amp;w=751&amp;h=423"
                  alt="Membership"
                />
              </picture>
            </figure>
          )}
          <div class="coop-c-signpost__content">
            <h3 class="coop-c-signpost__title">{title}</h3>
            <span class="coop-c-signpost__icon" aria-hidden="true">
              <svg class="coop-c-signpost__icon__svg" viewBox="0 0 16 29">
                <path d="M1.909 28.11a1.575 1.575 0 0 1-1.115-2.691L11.713 14.5.793 3.58a1.575 1.575 0 1 1 2.23-2.228l12.033 12.033a1.575 1.575 0 0 1 0 2.229L3.023 27.647c-.307.308-.71.463-1.114.463z" />
              </svg>
            </span>
          </div>
        </a>
      </div>
    </>
  );
};

Signpost.defaultProps = {
  title: null,
  url: null,
  mediaComponent: null,
};

Signpost.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  mediaComponent: PropTypes.string.node,
};

export default Signpost;
