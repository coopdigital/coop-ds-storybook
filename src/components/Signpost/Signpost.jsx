import React from "react";
import PropTypes from "prop-types";
import classNames from "../../utils/classNames";
import "./signpost.scss";

const Signpost = ({ title, url, mediaComponent }) => {
  const rootClasses = classNames("coop-c-signpost", [
    mediaComponent && "coop-c-hero--media",
  ]);

  return (
    <>
      <div className={rootClasses}>
        <a
          href={url}
          class="coop-c-signpost__link"
          data-contenttype=""
          data-contentparent=""
          data-linktext=""
        >
          { mediaComponent && ( mediaComponent ) }

          <div class="coop-c-signpost__content">
            <h3 class="coop-c-signpost__title">{title}</h3>
            <span class="coop-c-signpost__icon" aria-hidden="true">
              <svg class="coop-c-signpost__icon__svg" viewBox="0 0 16 29">
                <path d="M1.909 28.11a1.575 1.575 0 0 1-1.115-2.691L11.713 14.5.793 3.58a1.575 1.575 0 1 1 2.23-2.228l12.033 12.033a1.575 1.575 0 0 1 0 2.229L3.023 27.647c-.307.308-.71.463-1.114.463z" />
              </svg>
            </span>
          </div>
        </a>
      </div>
    </>
  );
};

Signpost.defaultProps = {
  title: "Supporting communities",
  url: "http://www.coop.co.uk",
  mediaComponent: null,
  dataContentType: null,
  dataContentParent: null,
  dataLinkText: null,
};

Signpost.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  mediaComponent: PropTypes.node,
  dataContentType: PropTypes.string.isOptional,
  dataContentParent: PropTypes.string.isOptional,
  dataLinkText: PropTypes.string.isOptional,
};

export default Signpost;
