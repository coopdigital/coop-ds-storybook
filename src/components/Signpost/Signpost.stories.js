import React from "react";
import Signpost from "./Signpost";
import ResponsiveImage from "../ResponsiveImage";

export default {
  title: "Base/Signpost",
  component: Signpost,
};

const linkTitle = "Supporting communities";
const url = `http://www.coop.co.uk`;

export const Basic = (args) => <Signpost {...args} />;
Basic.args = {
  title: linkTitle,
  url,
};

const MediaComponent = () => {
  const src =
    "https://images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=png&q=60&w=600&h=338";
  const alt = "This is the alt text";
  return <ResponsiveImage src={src} alt={alt} />;
};
export const WithMedia = () => (
  <Signpost title={linkTitle} url={url} mediaComponent={<MediaComponent />} />
);
