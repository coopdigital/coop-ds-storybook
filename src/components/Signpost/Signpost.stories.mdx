import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import Signpost from "./Signpost";

<Meta title="Base/Signpost" component={Signpost} />

# Signpost
A link card that can be displayed with or without an image.

Signpost navigation is used in pages to link to permanent content.

<code>
.a-class-name
</code>


<Story name="Signpost/All">
  <Signpost title="Signpost title" url="#" />
  <Signpost title="Signpost title" url="#" mediaComponent="true" />
</Story>
