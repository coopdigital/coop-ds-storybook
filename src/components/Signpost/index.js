import Signpost from "./Signpost";

export default Signpost;
