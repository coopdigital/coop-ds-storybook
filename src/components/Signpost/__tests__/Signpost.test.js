import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Signpost from "../index";
import ResponsiveImage from "../../ResponsiveImage";

const MediaComponent = () => {
  const src =
    "https://images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=png&q=60&w=600&h=338";
  const alt = "How Co-op is supporting local communities";
  const figureClass = "coop-c-signpost__media";
  const pictureClass = "coop-c-signpost__image";
  return (
    <ResponsiveImage
      src={src}
      alt={alt}
      figureClass={figureClass}
      pictureClass={pictureClass}
    />
  );
};

describe("Signpost", () => {
  it("should render correctly", () => {
    const wrapper = mount(<Signpost title="Sigpost title" url="#" />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer.create(<Signpost title="Sigpost title" url="#" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  //it should take a child component as mediacomponent without error
	it("should take a child component as mediacomponent", () => {
		const wrapper = mount(<Signpost title="Sigpost title" url="#" mediaComponent={<MediaComponent/>}/>);

		const tree = renderer.create(<Signpost />).toJSON();
    expect(tree).toMatchSnapshot();
	})

  //it should render the ImageComponent is used

  //it should carry a link

  //it should error without link or title
});
