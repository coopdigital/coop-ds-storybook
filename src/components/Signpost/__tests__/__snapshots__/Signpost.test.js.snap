// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signpost should render correctly 1`] = `
<div
  className="coop-c-signpost"
>
  <a
    className="coop-c-signpost__link"
    data-contentparent=""
    data-contenttype=""
    data-linktext=""
    href="#"
  >
    <div
      className="coop-c-signpost__content"
    >
      <h3
        className="coop-c-signpost__title"
      >
        Sigpost title
      </h3>
      <span
        aria-hidden="true"
        className="coop-c-signpost__icon"
      >
        <svg
          className="coop-c-signpost__icon__svg"
          viewBox="0 0 16 29"
        >
          <path
            d="M1.909 28.11a1.575 1.575 0 0 1-1.115-2.691L11.713 14.5.793 3.58a1.575 1.575 0 1 1 2.23-2.228l12.033 12.033a1.575 1.575 0 0 1 0 2.229L3.023 27.647c-.307.308-.71.463-1.114.463z"
          />
        </svg>
      </span>
    </div>
  </a>
</div>
`;

exports[`Signpost should take a child component as mediacomponent 1`] = `
<div
  className="coop-c-signpost"
>
  <a
    className="coop-c-signpost__link"
    data-contentparent=""
    data-contenttype=""
    data-linktext=""
    href="http://www.coop.co.uk"
  >
    <div
      className="coop-c-signpost__content"
    >
      <h3
        className="coop-c-signpost__title"
      >
        Supporting communities
      </h3>
      <span
        aria-hidden="true"
        className="coop-c-signpost__icon"
      >
        <svg
          className="coop-c-signpost__icon__svg"
          viewBox="0 0 16 29"
        >
          <path
            d="M1.909 28.11a1.575 1.575 0 0 1-1.115-2.691L11.713 14.5.793 3.58a1.575 1.575 0 1 1 2.23-2.228l12.033 12.033a1.575 1.575 0 0 1 0 2.229L3.023 27.647c-.307.308-.71.463-1.114.463z"
          />
        </svg>
      </span>
    </div>
  </a>
</div>
`;
